# Fixed-Point Iteration 

**TODO:** Come back and fill in some notes regarding fixed-point iteration, for now just implement the algorithm such that I can use it to work on the exercises from the homework/text.

In [1]:
import numpy as np
import math

In [2]:
# define g (from illustration on p.60 of text)
def g(x): 
    return ((1/2) * math.pow((10-x*x*x), (1/2)))

<br>**TODO:** Create an overloaded version of this function that takes multiple functions and evaluates them at and evaluate them at $p_0$ so that one can compare different versions of the function $g(x)$.

In [3]:
# overload this function, include iterations and error
def fixedPoint(p0, error, n, f):
    counter = 0
    while counter < n:
        p = f(p0)    # evaluate function at p0
        if math.fabs(p - p0) < error:    # reached error threshold
            return p
        else:
            counter = counter + 1
            p0 = p
            print(f"{counter:>02}\t{p0:<16}\t{p:<11}")
    print(f"The method failed after {n} iterations")

In [4]:
# overload function, iteration only 
def fixedPoint(p0, n, f):
    counter = 0
    print(f"n\tP")
    print("______________________________")
    while counter < n:
        p = f(p0)    # evaluate function at p0
        counter = counter + 1
        p0 = p
        print(f"{counter:>02}\t{p0:<16}")
    print(f"\nThe approximated value of P after {n} iterations is: {p}")

In [5]:
# test function using example from book
fixedPoint(1.5, 30, g)

n	P
______________________________
01	1.286953767623375
02	1.4025408035395783
03	1.3454583740232942
04	1.3751702528160383
05	1.360094192761733
06	1.3678469675921328
07	1.3638870038840212
08	1.36591673339004
09	1.364878217193677
10	1.365410061169957
11	1.3651378206692129
12	1.3652772085244786
13	1.3652058502970472
14	1.3652423837188388
15	1.3652236802252822
16	1.3652332557424998
17	1.365228353462627
18	1.3652308632436367
19	1.3652295783339587
20	1.3652302361581812
21	1.3652298993777325
22	1.3652300717962909
23	1.365229983524674
24	1.3652300287163228
25	1.36523000557995
26	1.3652300174248766
27	1.365230011360733
28	1.3652300144653395
29	1.3652300128759012
30	1.365230013689632

The approximated value of P after 30 iterations is: 1.365230013689632


---

## Exercise Set 2.2

This section implements solutions to some of the exercises and homework from the section 2.2 exercise set.

<br>
2. (a) Perform four iterations if possible on each of the functions $g$ defined in Exercise 1. 

Let $p_0=1$ and $p_{n+1}=g(p_n)$ for $n = 0,1,2,3$.

To start, let's provide a Python implementation for each of the four functions. 

* $g_1(x) = (3 + x - 2x^2)^{\frac{1}{4}}$

* $g_2(x) = \left(\frac{x+3-x^4}{2}\right)^\frac{1}{2}$

* $g_3(x) = \left(\frac{x+3}{x^2+2}\right)^\frac{1}{2}$

* $g_4(x) = \frac{3x^4+2x^2+3}{4x^3+4x-1}$

In [6]:
def g1(x):
    return math.pow(3+x-(2*x*x), (1/4))

In [7]:
def g2(x):
    return math.pow((x+3-math.pow(x,4))/2, (1/2))

In [8]:
def g3(x):
    return math.pow((x+3)/(x*x+2), (1/2))

In [9]:
def g4(x):
    return (3*math.pow(x,4) + 2*math.pow(x,2) + 3)/(4*math.pow(x,3) + 4*x - 1)

In [10]:
# compute four iterations with g1
fixedPoint(1, 4, g1)

n	P
______________________________
01	1.189207115002721
02	1.0800577526675623
03	1.1496714305893827
04	1.1078205295102599

The approximated value of P after 4 iterations is: 1.1078205295102599


In [11]:
# compute four iterations with g2
fixedPoint(1, 4, g2)

n	P
______________________________
01	1.224744871391589
02	0.9936661590774817
03	1.228568645274987
04	0.9875064291508866

The approximated value of P after 4 iterations is: 0.9875064291508866


In [12]:
fixedPoint(1, 4, g3)

n	P
______________________________
01	1.1547005383792515
02	1.116427409872122
03	1.1260522330022757
04	1.1236388847132548

The approximated value of P after 4 iterations is: 1.1236388847132548


In [13]:
fixedPoint(1, 4, g4)

n	P
______________________________
01	1.1428571428571428
02	1.1244816900178953
03	1.1241231639401488
04	1.124123029704334

The approximated value of P after 4 iterations is: 1.124123029704334


<br>
(b) which function do you think gives the best approximation to the solution?

<br>So, we want to figure out which one has the best approximation to the root for the function $f(x)=x^4+2x^2-x-3$?

Does this mean we need to use another method to compute the zero, if so, how do we go about it?

<br>First attempt, let's use the bisection method to find an approximation and then compare the approximation found with that to the approximations found using functions with fixed-point iteration. 

<br>**MAJOR TODO:** Start putting these functions in modules to easily import them instead of copy/paste when I need to call them.

In [14]:
# define our function f
def f(x):
    return (math.pow(x,4) + 2*math.pow(x,2) - x - 3)

In [15]:
# create function for bisection method algorithm
def bisection(a, b, t, n, func):
    counter = 0
    fa = func(a)
    p = a + (b-a)/2 
    l = a
    u = b
    print(f"n\ta\t\t\tb\t\t\tp (approx.)\t\t\tf(p)\t\t\t\tError")
    print("__________________________________________________________________________________________________________________________________________")
    while counter < n:
        p = a + (b-a)/2
        fp = func(p)
        if (fp == 0) or ((b-a)/2 < t):    # if exact root found or error tolerance is reached
            counter += 1
            print(f"{counter:>02}\t{a:<16}\t{b:<11}\t\t{p:<16}\t\t{func(p):<16}\t\t{b-p:<16}")
            break
        elif (fa * fp > 0):    # same signs, shift right
            print(f"{counter+1:>02}\t{a:<16}\t{b:<11}\t\t{p:<16}\t\t{func(p):<16}\t\t{b-p:<16}")
            a = p
            fa = fp
        else:    # different signs, shift left
            print(f"{counter+1:>02}\t{a:<16}\t{b:<11}\t\t{p:<16}\t\t{func(p):<16}\t\t{b-p:<16}")
            b = p
        counter += 1
    print(f"\nAfter {counter} iterations, the approximation for the root in [{l},{u}] is ~{p}\nwith error {b-p}")

<br>The graph of $g_1$ intersects $y=x$ at two different points, let's use the interval $[1,1.5]$ to try and get a "good" approximation for the root for $f$. 

In [16]:
bisection(1, 1.5, 0.00000001, 100, f)

n	a			b			p (approx.)			f(p)				Error
__________________________________________________________________________________________________________________________________________
01	1               	1.5        		1.25            		1.31640625      		0.25            
02	1               	1.25       		1.125           		0.008056640625  		0.125           
03	1               	1.125      		1.0625          		-0.5302581787109375		0.0625          
04	1.0625          	1.125      		1.09375         		-0.2700643539428711		0.03125         
05	1.09375         	1.125      		1.109375        		-0.1332949995994568		0.015625        
06	1.109375        	1.125      		1.1171875       		-0.06319832429289818		0.0078125       
07	1.1171875       	1.125      		1.12109375      		-0.027716427808627486		0.00390625      
08	1.12109375      	1.125      		1.123046875     		-0.009866390362731181		0.001953125     
09	1.123046875     	1.125      		1.1240234375    		-0.0009140116153503186		0.0009765625    
10	1.1240234375    	

<br>So, using the bisection method, $g_1$ appears to be converging to 1.124..., therefore, let's use this as a comparison for the results of our function.

Comparing the results from the bisection method with a very small error, one can see that the functions $g_3$ and $g_4$ provided good approximations of the actual root for our function $f$. However, I would say $g_4$ performed the best as it gave a very accurate approximation and converged to the approximation faster than $g_3$. It offered a good approximation after only 3 iterations.  

---

## Question 3

Use Fixed-Point Theorem to show 
$$g(x)=\frac{1}{10}\left(\frac{5}{x^2}+2x+9\right)$$
has a unique fixed point on $[1,3]$.

To go about this we need to use the theorem to show existence of a fixed-point in the interval and then we need to show uniqueness of the fixed-point in the interval. 

In [26]:
# implement g(x) as a Python function 
def g(x):
    return ((1/10) * ((5/(x*x))+2*x+9))

In [27]:
g(1)    # evaluate g(x) at first endpoint

1.6

In [28]:
g(3)

1.5555555555555556